docs/core/settings/variables.md

```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
+to CR LF. `%{user | substr(1)}` will take first character of username. Example of named parameters
+would be `%{user | md5(rounds=1000,salt='pepper')}`
+
+You can use `%%{variable}` to escape this, and emit `%{variable}`.
+
+Filters accept strings, numbers and variables as parameters. Parameters can be positional or
+named key-value pairs. Key names cannot be variables.
+The left side of pipe character (`|`) is provided as input to a filter. Some filters can be used
+in place of variables, e.g. lookup, literal and if.
+
+When value is missing or empty, you can use the `default` filter to provide value. Missing variables
+will cause errors and must be negated with default. This does not apply to all providers, some
+providers return empty when value is missing.
+
+If the last filter would output binary data, the data is encoded with `hexlify` filter by default.
+To avoid this, you can use `text` filter, which will sanitize the input and mark it as text.
+
+The new syntax also supports simple maths, you can do one operation. E.g. `%{port + 1000}`.
+Addition, substraction, multiplication, division and modulo operations are supported for now.
+
+A special case for modulo operation is that it can be applied to binary input, e.g. `md5 % 256`.
+The input is treated as 64-bit unsigned number and modulo is taken from that.
+
+All strings must be encapsulated with `"` or `'`, and you can escape them using `\\` within string.
+Numbers when used as parameters must be provided without quotes.
+
+## List of filters
+
+All parameters are strings unless stated otherwise.
+If parameter is `any`, it accepts both numbers and strings.
+Boolean type is `0` for false and `1` for true.
+
+Filters that have `any` in input and output mean that they will accept bytes or string, and the result
+will be bytes or strings, depending on the input.
+
+| Filter | Input | Output | Description |
+| ------ | ----- | ------ | ----------- |
+| base64(pad=boolean, url=boolean) | String | String | Base64 encode given input, defaults to pad and not url scheme. |
+| benumber | Byte | String | Convert big-endian encoded input into a number. |
+| concat(any, any...) | Byte | Byte | Concatenates input with value(s). Numbers are coerced to strings. Input is optional. |
+| default(value) | String | String | Replace empty or missing input with value. Clears missing variable error. If no value is provided, empty string is used. |
+| domain | String | String | Provides domain part of user@domain value. |
+| hash(method, rounds=number, salt=string) | Byte | Byte | Returns raw hash from input using given hash method. Rounds and salt are optional. |
+| hexlify(width) | Byte | String | Convert bytes into hex with optional width, truncates or pads up to width. |
+| hex(width) | Number | String | Convert base-10 number to base-16 number. If width is specified the result is truncated or padded with 0 to width. Negative width is applied after number. |
```

Comment:
This is the input filter that uses Number as input/output. That doesn't seem right. Shouldn't at least benumber, lenumber have Number output?